modules/bullet/collision_object_bullet.cpp

```diff
@@ -329,59 +354,71 @@ bool RigidCollisionObjectBullet::is_shape_disabled(int p_index) {
 	return !shapes[p_index].active;
 }
 
+void RigidCollisionObjectBullet::pre_process() {
+	if (need_shape_reload) {
+		do_reload_shapes();
+		need_shape_reload = false;
+	}
+	CollisionObjectBullet::pre_process();
+}
+
 void RigidCollisionObjectBullet::shape_changed(int p_shape_index) {
 	ShapeWrapper &shp = shapes.write[p_shape_index];
 	if (shp.bt_shape == mainShape) {
 		mainShape = NULL;
 	}
-	bulletdelete(shp.bt_shape);
+	shp.release_bt_shape();
 	reload_shapes();
 }
 
 void RigidCollisionObjectBullet::reload_shapes() {
+	need_shape_reload = true;
+	if (space) {
+		space->add_to_flush_queue(this);
+	}
```

Comment:
The shape should not be reloaded during the flush but during the pre_process. To fix the issue #40840 you need to insert `p_body->pre_process();` at the beginning of the function `test_body_motion`.

Reply:
I don't understand what you mean. I already added `p_body->pre_process();` in `test_body_motion` on the line you mentioned in this comment https://github.com/godotengine/godot/pull/40886#issuecomment-666964762 (see https://github.com/godotengine/godot/pull/40188/commits/24d50fde7bb5bc528cb916a4b7d48fced7adff10#diff-a61d01bcaef4874b6713c0b00e6d4571R1014)